### PR TITLE
New payments design

### PR DESCRIPTION
**Pull request by Whisno**
- Removed payment wizard used in invoices
- Added account.payment model to use instead, with or without invoice
- Added account.payment.method model and use it on account_journal with fields inbound_payment_methods and outbound_payment_methods
- Added transfer_account on company

### CODE REVIEW

File: addons/account/account.py [hunk 1]
```diff
@@ -275,11 +275,25 @@ class AccountJournal(models.Model):
...
+        domain=[('type', '=', 'outbound')], string='Outbound Payment Methods', default=lambda self: [self.env.ref('account.account_payment_method_manual_out').id])
+
     _sql_constraints = [
         ('code_company_uniq', 'unique (code, name, company_id)', 'The code and name of the journal must be unique per company !'),
     ]
 
     @api.one
+    @api.constrains('inbound_payment_methods', 'outbound_payment_methods')
+    def _check_payment_methods(self):
+        # TODO: or allow to set no payment method in which case the journal can't be used for the corresponding payment operation ?
```
Comment (qdp-odoo) [inline addons/account/account.py]: remove this constraint

File: addons/account/account_payment.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+# TODO: multicurrency and multicompany
+
+class account_payment_method(models.Model):
+    _name = "account.payment.method"
+    _description = "Payment Methods"
+
+    name = fields.Char(required=True)
+    code = fields.Char(required=True) # For internal identification
+    type = fields.Selection([('inbound', 'Inbound'), ('outbound', 'Outbound')], required=True)
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: use something else than 'type' as variable name

File: addons/account/account.py (hunk 1 above)
Comment (Whisno) [inline addons/account/account.py]: so a journal with no inbound / outbound payment method simply cannot be used for inbound / outbound payments ?

Comment (qdp-odoo) [inline addons/account/account.py]: yep

File: addons/account/account_payment.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    def _compute_payment_difference(self):
+        self.payment_difference = self.invoice_id and self.invoice_id.residual - self.amount or 0
+
+    name = fields.Char(readonly=True, copy=False)
+    state = fields.Selection([('draft','Draft'), ('confirmed','Confirmed')], readonly=True, default='draft')
+
+    payment_type = fields.Selection([('outbound', 'Send Money'), ('inbound', 'Receive Money'), ('transfer', 'Transfer Money')], default='outbound', required=True)
+    payment_method = fields.Many2one('account.payment.method', string='Payment Method', required=True)
+    # default True to prevent a visual glitch
+    hide_payment_method = fields.Boolean(default=True, copy=False)
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: remove hide_payment_method

File: addons/account/account_payment.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+            'account_id': self.destination_account_id.id,
+            'journal_id': self.journal_id.id,
+            'payment_id': self.id,
+        }
+
+    def _get_liquidity_move_line_vals(self):
+        if self.payment_type == 'transfer':
+            name = 'TODO'
+        else:
+            prefixes = [["Payment to ", "Refund from "],
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: _("translate_me")

refactor this to be less ugly

File: addons/account/account_payment.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        return {
+            'name': name,
+            'account_id': self.payment_type == 'outbound' \
+                and self.journal_id.default_debit_account_id.id \
+                or self.journal_id.default_credit_account_id.id,
+            'journal_id': self.journal_id.id,
+        }
+
+    @api.model
+    def create(self, vals):
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: put in pay_and_reconcile

File: addons/account/account_payment.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        name = journal.sequence_id.next_by_id()
+        return {
+            'name': name,
+            'date': self.date,
+            'ref': self.reference or name,
+            'company_id': self.company_id.id,
+            'journal_id': journal.id,
+        }
+
+    def _get_shared_move_line_vals(self, debit, credit, amount_currency, move_id):
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: define this method in post() as it's the only place it's used

File: addons/account/account_payment.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+            'date': self.date,
+            'invoice': self.invoice_id and self.invoice_id.id or False,
+            'move_id': move_id,
+            'debit': debit,
+            'credit': credit,
+            'currency_id': self.journal_id.currency.id,
+            'amount_currency': amount_currency or False,
+        }
+
+    def _get_counterpart_move_line_vals(self):
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: rm

File: addons/account/account_payment.py [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+        else:
+            name = 'TODO'
+        return {
+            'name': name,
+            'account_id': self.destination_account_id.id,
+            'journal_id': self.journal_id.id,
+            'payment_id': self.id,
+        }
+
+    def _get_liquidity_move_line_vals(self):
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: rm

File: addons/account/account_payment.py [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+            # Reconcile the invoice, if present
+            if res.invoice_id:
+                if res.payment_difference != 0.0 and res.payment_difference_handling == 'reconcile':
+                    res.invoice_id.register_payment(counterpart_aml, res.writeoff_account, res.journal_id)
+                else:
+                    res.invoice_id.register_payment(counterpart_aml)
+
+            # In case of a transfer, the first journal entry created debited the source liquidity account and credited
+            # the transfer account. Now we debit the transfer account and credit the destination liquidity account.
+            if res.payment_type == 'transfer':
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: refactor

File: addons/account/account_payment.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+                dst_move.post()
+
+                (counterpart_aml + transfer_debit_aml).reconcile()
+
+            res.state = 'confirmed'
+            if res.payment_method.code == 'manual':
+                res.payment_state = 'done'
+
+    @api.multi
+    def button_create_and_post(self):
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: remove

File: addons/account/account_payment.py [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+                self.invalidate_cache()
+                self.env['account.move'].invalidate_cache()
+                self.env['account.move.line'].invalidate_cache()
+
+                rec.unlink()
+                self._cr.commit()
+                raise
+
+    @api.multi
+    def unlink(self):
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: replace method name by button_cancel(),  unlink is allowed only when state=='canceled'

File: addons/account/company.py [hunk 12]
```diff
@@ -10,6 +10,9 @@ class ResCompany(models.Model):
     fiscalyear_last_month = fields.Selection([(1, 'January'), (2, 'February'), (3, 'March'), (4, 'April'), (5, 'May'), (6, 'June'), (7, 'July'), (8, 'August'), (9, 'September'), (10, 'October'), (11, 'November'), (12, 'December')], default=12)
     period_lock_date = fields.Date(help="Only users with the 'Adviser' role can edit accounts prior to and inclusive of this date")
     fiscalyear_lock_date = fields.Date(string="Fiscal Year lock date", help="No users, including Advisers, can edit accounts prior to and inclusive of this date")
+    # TODO: Set this via l10n_xx muse have reconcile = True and user_type = self.env.ref('account.data_account_type_current_assets')
```
Comment (qdp-odoo) [inline addons/account/company.py]: you can do it for l10n_be

File: addons/account/demo/account_minimal.xml [hunk 13]
```diff
@@ -288,5 +288,17 @@
...
+
+        <!-- TODO: set company_id.transfer_account via l10n_xx -->
+        <record id="trans" model="account.account">
+            <field name="code">X1019</field>
+            <field name="reconcile" eval='True'/>
+            <field name="name">Liquidity Transfers - (test)</field>
+            <field name="user_type" ref="data_account_type_current_assets"/>
+        </record>
+        <record id="base.main_company" model="res.company">
+            <field name="transfer_account" ref="trans"/>
```
Comment (qdp-odoo) [inline addons/account/demo/account_minimal.xml]: transfer_account_id

File: addons/account/views/account_view.xml [hunk 14]
```diff
@@ -273,6 +273,8 @@
                                     <field name="show_on_dashboard"/>
                                 </group>
                                 <group>
+                                    <field name="inbound_payment_methods" create="false" widget="many2many_checkboxes" attrs="{'invisible': [('type', 'not in', ('bank', 'cash'))]}"/>
```
Comment (qdp-odoo) [inline addons/account/views/account_view.xml]: just bank

File: addons/account/account_payment.py [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+    move_lines = fields.One2many('account.move.line', 'payment_id', copy=False, readonly=True, ondelete='restrict')
+
+    @api.one
+    @api.constrains('amount')
+    def _check_amount(self):
+        if self.amount == 0.0:
+            raise ValidationError('Please enter payment amount.')
+        if self.amount < 0.0:
+            raise ValidationError('The payment amount must always be positive.')
+        if self.invoice_id and self.amount > self.invoice_id.residual:
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: remove this test

File: addons/account/account_payment.py [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+    payment_difference = fields.Float(compute='_compute_payment_difference')
+    payment_difference_handling = fields.Selection([('open', 'Keep Open'), ('reconcile', 'Reconcile Payment Balance')], default='open', string="Payment Difference")
+    writeoff_account = fields.Many2one('account.account', string="Counterpart Account", domain=[('deprecated', '=', False)])
+
+    move_lines = fields.One2many('account.move.line', 'payment_id', copy=False, readonly=True, ondelete='restrict')
+
+    @api.one
+    @api.constrains('amount')
+    def _check_amount(self):
+        if self.amount == 0.0:
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: self.amont > 0:
  raise ValidationError('The payment amount must be strictly positive.')

File: addons/account/views/account_payment_view.xml [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+                    <header>
+                        <button name="post" class="oe_highlight" states="draft" string="Confirm" type="object"/>
+                        <field name="state" widget="statusbar"/>
+                    </header>
+                    <sheet>
+                        <group>
+                            <group>
+                                <field name="payment_type" widget="radio" attrs="{'readonly': [('state', '=', 'confirmed')]}"/>
+                                <field name="partner_type" widget="radio" attrs="{'required': [('payment_type', '!=', 'transfer')], 'invisible': [('payment_type', 'in', ('transfer', False))], 'readonly': [('state', '=', 'confirmed')]}"/>
+                                <field name="invoice_id" attrs="{'invisible': ['|', ('payment_type', '=', 'transfer'), '&amp;', ('state', '=', 'confirmed'), ('invoice_id', '=', False)], 'readonly': [('state', '=', 'confirmed')]}"/>
```
Comment (qdp-odoo) [inline addons/account/views/account_payment_view.xml]: invoice_id is always invisible => that means the onchanges are much easier

File: addons/account/account_payment.py [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+            if payment_methods:
+                self.payment_method = payment_methods[0]
+            # Hide payment method if only 'Manual' is available
+            self.hide_payment_method = len(payment_methods) == 1 and payment_methods[0].code == 'manual'
+            # Set payment method domain (restrict to methods enabled for the journal and to selected payment type)
+            payment_type = self.payment_type in ('outbound', 'transfer') and 'outbound' or 'inbound'
+            return {'domain': {'payment_method': [('type', '=', payment_type), ('id', 'in', payment_methods.ids)]}}
+
+    @api.model
+    def default_get(self, fields):
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: remove default_get and put 'res.update({'invoice_id': self._context['invoice_id']})' in the invoice_id field definition

File: addons/account/account_payment.py [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+            # self.invoice_id = False resulting in a buggy behaviour.
+            res.pop('payment_type', None)
+            res.pop('partner_type', None)
+        return res
+
+    def _get_move_vals(self, journal=None):
+        """ Return dict to create the payment move """
+        journal = journal or self.journal_id
+        if not journal.sequence_id:
+            raise UserError(_('Configuration Error !'), _('The journal ' + journal.name + ' does not have a sequence, please specify one.'))
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: this:
   ```_('The journal ' + journal.name + ' does not have a sequence, please specify one.'))```
is not a valid string to translate. You must use
   ```_('The journal %s does not have a sequence, please specify one.') % journal.name)```

File: addons/account/account_payment.py [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+        if not journal.sequence_id.active:
+            raise UserError(_('Configuration Error !'), _('The sequence of journal ' + journal.name + ' is deactivated.'))
+        name = journal.sequence_id.next_by_id()
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: add in context: ir_sequence_date =  self.date when calling next_by_id()

File: addons/account/account_payment.py [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+            'date': self.date,
+            'ref': self.reference or name,
+            'company_id': self.company_id.id,
+            'journal_id': journal.id,
+        }
+
+    def _get_shared_move_line_vals(self, debit, credit, amount_currency, move_id):
+        """ Returns values common to both move lines (except for debit, credit and amount_currency which are reversed) """
+        return {
+            'partner_id': self.payment_type in ('inbound', 'outbound') and self.partner_id.id or False, # TOCHECK .commercial_partner_id.id ?
```
Comment (qdp-odoo) [inline addons/account/account_payment.py]: commercial_partner_id

File: addons/account/account.py (hunk 1 above)
Comment (Whisno) [inline addons/account/account.py]: done

File: addons/account/account_payment.py (hunk 2 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/account_payment.py (hunk 3 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/account_payment.py (hunk 5 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/account_payment.py (hunk 10 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/account_payment.py (hunk 11 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/views/account_view.xml (hunk 14 above)
Comment (Whisno) [inline addons/account/views/account_view.xml]: done

File: addons/account/account_payment.py (hunk 15 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/account_payment.py (hunk 16 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/views/account_payment_view.xml (hunk 17 above)
Comment (Whisno) [inline addons/account/views/account_payment_view.xml]: done

File: addons/account/account_payment.py (hunk 18 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/account_payment.py (hunk 19 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/account_payment.py (hunk 20 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

File: addons/account/account_payment.py (hunk 21 above)
Comment (Whisno) [inline addons/account/account_payment.py]: done

**Pull request merged**